cmd/autoheal/server.go

```diff
@@ -81,25 +81,28 @@ func serverRun(cmd *cobra.Command, args []string) {
 	// Load the Kubernetes configuration:
 	var config *rest.Config
 	_, err = os.Stat(serverKubeConfig)
-	if os.IsNotExist(err) {
-		glog.Infof(
-			"The Kubernetes configuration file '%s' doesn't exist, will try to use the "+
-				"in-cluster configuration",
-			serverKubeConfig,
-		)
-		config, err = rest.InClusterConfig()
+	file, err := os.Open(serverKubeConfig) // For read access.
+
+	if err == nil {
+		file.Close()
```

Comment:
defer file.Close() instead?

Reply:
A way I've seen used to detect running under kubernetes is checking for `KUBERNETES_SERVICE_HOST` env var.
This var (plus _PORT) is currently what the `InClusterConfig` in the client uses to find the API:
https://github.com/kubernetes/client-go/blob/989be4278f353e42f26c416c53757d16fcff77db/rest/config.go#L313-L314

As discussed on https://github.com/kubernetes/kubernetes/issues/40973, this var has 2 problems:
1. It can be shadowed by a service that happens to be called `kubernetes` in current namespace.  (Not a problem for us, we control what's running in our namespace.)
2. it's sometimes unset due to startup race condition.  This being a tool for healing problems, we don't want to fail when problems happen :-)  A possibly better approach to find API is DNS `kubernetes.default.svc`; for now official clients were not changed (https://github.com/kubernetes/kubernetes/pull/43807 was rejected), but we can opt-in by explicitly setting `KUBERNETES_SERVICE_HOST` to `kubernetes.default.svc` in the pod config.

In any case, for the decision we can rely on whatever InClusterConfig will doing.  (If we're in a pod and it fails, we're hosed anyway.  But if we exit with error, kubelet can restart us, with backoff, which is not bad.)
=> I propose trying InClusterConfig first — it returns error if it doesn't find things it expects to have in a pod! — and then falling back to `~/.kube/config`.